### PR TITLE
compute environment sha unique

### PR DESCRIPTION
**Pull request by evgeniilezhnin**

### CODE REVIEW

File: backend/src/main/java/ai/verta/modeldb/versioning/blob/container/EnvironmentContainer.java [hunk 1]
```diff
@@ -208,7 +208,8 @@ private void validateEnvironmentVariableName(String name) throws ModelDBExceptio
   private String computeSHA(EnvironmentBlob blob) throws NoSuchAlgorithmException {
     StringBuilder sb = new StringBuilder();
     sb.append("env:");
-    for (EnvironmentVariablesBlob environmentVariablesBlob : blob.getEnvironmentVariablesList()) {
+    for (EnvironmentVariablesBlob environmentVariablesBlob :
+        new HashSet<>(blob.getEnvironmentVariablesList())) {
```
Comment (anandjakhaniya) [inline backend/src/main/java/ai/verta/modeldb/versioning/blob/container/EnvironmentContainer.java]: Why this change required? it just for loop you can use whatever I mean List or Set I think it doesn't affect to hash also.

**Comment by ravishetye** [comment]
@evgeniilezhnin  this can be closed? or still, needs review?

File: backend/src/main/java/ai/verta/modeldb/versioning/blob/container/EnvironmentContainer.java (hunk 1 above)
Comment (evgeniilezhnin) [reply backend/src/main/java/ai/verta/modeldb/versioning/blob/container/EnvironmentContainer.java]: HashSet ignores order of elements. That's why I have used it. But it doesn't guarantee order, so I better change it to sort.

**Comment by evgeniilezhnin** [comment]
> @evgeniilezhnin this can be closed? or still, needs review?

it needs review

File: backend/src/main/java/ai/verta/modeldb/versioning/blob/container/EnvironmentContainer.java [hunk 2]
```diff
@@ -227,23 +222,9 @@ private String computeSHA(PythonEnvironmentBlob blob) throws NoSuchAlgorithmExce
...
     sb.append(":requirements:");
-    for (PythonRequirementEnvironmentBlob pythonRequirementEnvironmentBlob :
-        blob.getRequirementsList()) {
-      sb.append(":library:")
-          .append(pythonRequirementEnvironmentBlob.getLibrary())
-          .append(":constraint:")
-          .append(pythonRequirementEnvironmentBlob.getConstraint());
-      appendVersion(sb, pythonRequirementEnvironmentBlob.getVersion());
-    }
+    sb.append(new HashSet<>(blob.getRequirementsList()).hashCode());
```
Comment (ravishetye) [inline backend/src/main/java/ai/verta/modeldb/versioning/blob/container/EnvironmentContainer.java]: would hahsCode be same across calls/ across restarts?
we would want the same content of requirementList to be create identical sha

**Comment by conradoverta** [comment]
The autogenerated code has hash computation, which should be stable (it checks for new fields so that we can add them without changing the SHA). If we want to change how SHA is computed, should we use that instead?

File: backend/src/main/java/ai/verta/modeldb/versioning/blob/container/EnvironmentContainer.java (hunk 2 above)
Comment (evgeniilezhnin) [reply backend/src/main/java/ai/verta/modeldb/versioning/blob/container/EnvironmentContainer.java]: changed base and used generated files to compute hash

File: backend/src/main/java/ai/verta/modeldb/versioning/blob/container/EnvironmentContainer.java [hunk 3]
```diff
@@ -208,12 +208,15 @@ private void validateEnvironmentVariableName(String name) throws ModelDBExceptio
...
-    }
+    List<
+            ai.verta.modeldb.versioning.autogenerated._public.modeldb.versioning.model
+                .EnvironmentVariablesBlob>
+        environmentVariables =
+            ai.verta.modeldb.versioning.autogenerated._public.modeldb.versioning.model
+                .EnvironmentBlob.fromProto(blob)
+                .EnvironmentVariables;
+    sb.append(
+        environmentVariables == null ? "null" : new HashSet<>(environmentVariables).hashCode());
```
Comment (anandjakhaniya) [inline backend/src/main/java/ai/verta/modeldb/versioning/blob/container/EnvironmentContainer.java]: why you are set to null as a string in StringBuilder? here something wrong. you should just leave it if it is null.

File: backend/src/main/java/ai/verta/modeldb/versioning/blob/container/EnvironmentContainer.java [hunk 4]
```diff
@@ -226,24 +229,15 @@ private String computeSHA(PythonEnvironmentBlob blob) throws NoSuchAlgorithmExce
...
-    for (PythonRequirementEnvironmentBlob pythonRequirementEnvironmentBlob :
-        blob.getRequirementsList()) {
-      sb.append(":library:")
-          .append(pythonRequirementEnvironmentBlob.getLibrary())
-          .append(":constraint:")
-          .append(pythonRequirementEnvironmentBlob.getConstraint());
-      appendVersion(sb, pythonRequirementEnvironmentBlob.getVersion());
-    }
+    sb.append(
+        blobNew.Requirements == null ? "null" : new HashSet<>(blobNew.Requirements).hashCode());
```
Comment (anandjakhaniya) [inline backend/src/main/java/ai/verta/modeldb/versioning/blob/container/EnvironmentContainer.java]: Same comment like above

File: backend/src/main/java/ai/verta/modeldb/versioning/blob/container/EnvironmentContainer.java [hunk 5]
```diff
@@ -226,24 +229,15 @@ private String computeSHA(PythonEnvironmentBlob blob) throws NoSuchAlgorithmExce
...
     sb.append(":constraints:");
-    for (PythonRequirementEnvironmentBlob pythonConstraintEnvironmentBlob :
-        blob.getConstraintsList()) {
-      sb.append(":library:")
-          .append(pythonConstraintEnvironmentBlob.getLibrary())
-          .append(":constraint:")
-          .append(pythonConstraintEnvironmentBlob.getConstraint());
-      appendVersion(sb, pythonConstraintEnvironmentBlob.getVersion());
-    }
+    sb.append(blobNew.Constraints == null ? "null" : new HashSet<>(blobNew.Constraints).hashCode());
```
Comment (anandjakhaniya) [inline backend/src/main/java/ai/verta/modeldb/versioning/blob/container/EnvironmentContainer.java]: same as above comments

File: backend/src/main/java/ai/verta/modeldb/versioning/blob/container/EnvironmentContainer.java (hunk 4 above)
Comment (evgeniilezhnin) [reply backend/src/main/java/ai/verta/modeldb/versioning/blob/container/EnvironmentContainer.java]: fixed

File: backend/src/main/java/ai/verta/modeldb/versioning/blob/container/EnvironmentContainer.java (hunk 3 above)
Comment (evgeniilezhnin) [reply backend/src/main/java/ai/verta/modeldb/versioning/blob/container/EnvironmentContainer.java]: fixed

File: backend/src/main/java/ai/verta/modeldb/versioning/blob/container/EnvironmentContainer.java (hunk 5 above)
Comment (evgeniilezhnin) [reply backend/src/main/java/ai/verta/modeldb/versioning/blob/container/EnvironmentContainer.java]: fixed

File: backend/src/main/java/ai/verta/modeldb/versioning/blob/container/EnvironmentContainer.java [hunk 6]
```diff
@@ -220,12 +220,16 @@ private void validateEnvironmentVariableName(String name) throws ModelDBExceptio
...
+    List<
+            ai.verta.modeldb.versioning.autogenerated._public.modeldb.versioning.model
+                .EnvironmentVariablesBlob>
+        environmentVariables =
+            ai.verta.modeldb.versioning.autogenerated._public.modeldb.versioning.model
+                .EnvironmentBlob.fromProto(blob)
+                .getEnvironmentVariables();
+    if (environmentVariables != null) {
+      sb.append("env:");
+      sb.append(new HashSet<>(environmentVariables).hashCode());
```
Comment (ravishetye) [inline backend/src/main/java/ai/verta/modeldb/versioning/blob/container/EnvironmentContainer.java]: Are you sure Hash set will guarantee you the same order across restarts? We want the same set of envVars to map to same SHA.

Why not use environmentVariables itself? It is getting sorted in the auto gen code.

File: backend/src/main/java/ai/verta/modeldb/versioning/blob/container/EnvironmentContainer.java [hunk 7]
```diff
@@ -239,23 +243,21 @@ private String computeSHA(PythonEnvironmentBlob blob) throws NoSuchAlgorithmExce
...
-      appendVersion(sb, pythonRequirementEnvironmentBlob.getVersion());
+    ai.verta.modeldb.versioning.autogenerated._public.modeldb.versioning.model.PythonEnvironmentBlob
+        blobNew =
+            ai.verta.modeldb.versioning.autogenerated._public.modeldb.versioning.model
+                .PythonEnvironmentBlob.fromProto(blob);
+    List<ai.verta.modeldb.versioning.autogenerated._public.modeldb.versioning.model.PythonRequirementEnvironmentBlob> requirements = blobNew
+        .getRequirements();
+    if (requirements != null) {
+      sb.append(":requirements:");
+      sb.append(new HashSet<>(requirements).hashCode());
```
Comment (ravishetye) [inline backend/src/main/java/ai/verta/modeldb/versioning/blob/container/EnvironmentContainer.java]: similar to above.

File: backend/src/main/java/ai/verta/modeldb/versioning/blob/container/EnvironmentContainer.java [hunk 8]
```diff
@@ -239,23 +243,21 @@ private String computeSHA(PythonEnvironmentBlob blob) throws NoSuchAlgorithmExce
...
-      sb.append(":library:")
-          .append(pythonConstraintEnvironmentBlob.getLibrary())
-          .append(":constraint:")
-          .append(pythonConstraintEnvironmentBlob.getConstraint());
-      appendVersion(sb, pythonConstraintEnvironmentBlob.getVersion());
+    List<ai.verta.modeldb.versioning.autogenerated._public.modeldb.versioning.model.PythonRequirementEnvironmentBlob> constraints = blobNew
+        .getConstraints();
+    if (constraints != null) {
+      sb.append(":constraints:");
+      sb.append(new HashSet<>(constraints).hashCode());
```
Comment (ravishetye) [inline backend/src/main/java/ai/verta/modeldb/versioning/blob/container/EnvironmentContainer.java]: same.

**Review by ravishetye — COMMENTED** [review]
I do not get the requirement of hashSet here

File: backend/src/main/java/ai/verta/modeldb/versioning/blob/container/EnvironmentContainer.java (hunk 6 above)
Comment (evgeniilezhnin) [reply backend/src/main/java/ai/verta/modeldb/versioning/blob/container/EnvironmentContainer.java]: oh sorry. Forgot about that. Fixed it. Also added a small change into template to use setters instead of set operator to force sort in fromProto.

File: backend/src/main/java/ai/verta/modeldb/versioning/blob/container/EnvironmentContainer.java (hunk 7 above)
Comment (evgeniilezhnin) [reply backend/src/main/java/ai/verta/modeldb/versioning/blob/container/EnvironmentContainer.java]: fixed

File: backend/src/main/java/ai/verta/modeldb/versioning/blob/container/EnvironmentContainer.java (hunk 8 above)
Comment (evgeniilezhnin) [reply backend/src/main/java/ai/verta/modeldb/versioning/blob/container/EnvironmentContainer.java]: fixed

File: backend/src/main/java/ai/verta/modeldb/versioning/blob/container/EnvironmentContainer.java (hunk 7 above)
Comment (ravishetye) [reply backend/src/main/java/ai/verta/modeldb/versioning/blob/container/EnvironmentContainer.java]: shouldn't this be removed?

File: backend/src/main/java/ai/verta/modeldb/versioning/blob/visitors/Visitor.java [hunk 9]
```diff
@@ -3,101 +3,103 @@
...
 import java.util.List;
+import java.util.stream.Collectors;
 
 public class Visitor {
-  public void preVisitListOfAutogenBlob(List<AutogenBlob> lst) throws ModelDBException {
+  public void preVisitListOfAutogenBlobDiff(List<AutogenBlobDiff> lst) throws ModelDBException {
     if (lst == null) return;
-    for (AutogenBlob val : lst) {
-      preVisitAutogenBlob(val);
+    for (AutogenBlobDiff val : lst) {
```
Comment (ravishetye) [inline backend/src/main/java/ai/verta/modeldb/versioning/blob/visitors/Visitor.java]: I do not understand why this file changed

File: backend/src/test/java/ai/verta/modeldb/blobs/BlobEquality.java [hunk 10]
```diff
@@ -1,22 +1,29 @@
 package ai.verta.modeldb.blobs;
 
-import static org.junit.Assert.*;
-import static org.junit.Assume.*;
-
+import ai.verta.modeldb.ModelDBException;
```
Comment (ravishetye) [inline backend/src/test/java/ai/verta/modeldb/blobs/BlobEquality.java]: i do not understand why this file changed

File: backend/src/main/java/ai/verta/modeldb/versioning/blob/visitors/Visitor.java (hunk 9 above)
Comment (conradoverta) [reply backend/src/main/java/ai/verta/modeldb/versioning/blob/visitors/Visitor.java]: I believe this file is auto-generated (it might be missing the header).

Comment (evgeniilezhnin) [reply backend/src/main/java/ai/verta/modeldb/versioning/blob/visitors/Visitor.java]: for some reason I have different result than in master when I run ./swagger_codegen.sh and reformat step in maven. Did I miss something here?

Comment (evgeniilezhnin) [reply backend/src/main/java/ai/verta/modeldb/versioning/blob/visitors/Visitor.java]: It mostly imports and method order.

I did just a small change in generation template. But when I don't do any changes it's still different

Comment (ravishetye) [reply backend/src/main/java/ai/verta/modeldb/versioning/blob/visitors/Visitor.java]: @conradoverta : could you review the autogen files and the process. the code merged from #425 makes sense to me otherwise.

**Review by ravishetye — APPROVED** [review]
